missing token on refresh

The following error reported 
```
Traceback (most recent call last):
  File "/opt/dev/miniconda3/envs/sm38/lib/python3.8/site-packages/lithops/serverless/backends/code_engine/code_engine.py", line 72, in __init__
    config.load_kube_config(config_file=self.kubecfg_path)
  File "/opt/dev/miniconda3/envs/sm38/lib/python3.8/site-packages/kubernetes/config/kube_config.py", line 798, in load_kube_config
    loader.load_and_set(config)
  File "/opt/dev/miniconda3/envs/sm38/lib/python3.8/site-packages/kubernetes/config/kube_config.py", line 576, in load_and_set
    self._load_authentication()
  File "/opt/dev/miniconda3/envs/sm38/lib/python3.8/site-packages/kubernetes/config/kube_config.py", line 280, in _load_authentication
    if self._load_auth_provider_token():
  File "/opt/dev/miniconda3/envs/sm38/lib/python3.8/site-packages/kubernetes/config/kube_config.py", line 299, in _load_auth_provider_token
    return self._load_oid_token(provider)
  File "/opt/dev/miniconda3/envs/sm38/lib/python3.8/site-packages/kubernetes/config/kube_config.py", line 403, in _load_oid_token
    self._refresh_oidc(provider)
  File "/opt/dev/miniconda3/envs/sm38/lib/python3.8/site-packages/kubernetes/config/kube_config.py", line 459, in _refresh_oidc
    refresh = request.refresh_token(
  File "/opt/dev/miniconda3/envs/sm38/lib/python3.8/site-packages/requests_oauthlib/oauth2_session.py", line 446, in refresh_token
    self.token = self._client.parse_request_body_response(r.text, scope=self.scope)
  File "/opt/dev/miniconda3/envs/sm38/lib/python3.8/site-packages/oauthlib/oauth2/rfc6749/clients/base.py", line 421, in parse_request_body_response
    self.token = parse_token_response(body, scope=scope)
  File "/opt/dev/miniconda3/envs/sm38/lib/python3.8/site-packages/oauthlib/oauth2/rfc6749/parameters.py", line 431, in parse_token_response
    validate_token_parameters(params)
oauthlib.oauth2.rfc6749.errors.MissingTokenError: (missing_token) Missing access token parameter.
```

Lithops version 2.4.1. Occurs from time to time, probably every 5-10 launches.
@JosepSampe any ideas?

@JosepSampe I tested yesterday. One of the launches got a MemoryError:
```
Activation: 'metaspace2020/metaspace-lithops-ce:1.8.6.9' (393e5cfad1c4)
[
    2021-09-06 14:53:06,333 [INFO] lithops.worker.handler -- Lithops v2.4.1 - Starting Code Engine (Job) execution
    2021-09-06 14:53:06,333 [INFO] lithops.worker.handler -- Execution ID: 367861-10-M007/00000
    2021-09-06 14:53:07,229 [INFO] lithops.worker.jobrunner -- Going to execute 'save_png_chunk()'
    ---------------------- FUNCTION LOG ----------------------
    2021-09-06 14:58:26,828 [ERROR] lithops.worker.handler -- No completion message received from JobRunner process
    ----------------------- EXCEPTION !-----------------------
    Traceback (most recent call last):
      File "/lithops/lithops/worker/handler.py", line 183, in run_job
        raise MemoryError('HANDLER', msg)
    MemoryError: ('HANDLER', 'Function exceeded maximum memory and was killed')
    ----------------------------------------------------------
]
```
However, ConfigMap `lithops-72f2b4-6-m008-configmap` was not deleted:
```
$ ibmcloud ce configmap list
Listing configmaps...
OK

Name                              Data  Age  

lithops-367861-10-m007-configmap  1     18h  
```

--- a/lithops/serverless/backends/code_engine/code_engine.py
+++ b/lithops/serverless/backends/code_engine/code_engine.py
@@ -58,13 +58,11 @@ class CodeEngineBackend:
         self.namespace = code_engine_config.get('namespace', None)
         self.region = code_engine_config.get('region', None)
 
+        self.ibm_token_manager = None
+
         if self.namespace and self.region and self.iam_api_key:
             self.cluster = ce_config.CLUSTER_URL.format(self.region)
-            configuration = client.Configuration()
-            configuration.host = self.cluster
-            token = self._get_iam_token()
-            configuration.api_key = {"authorization": "Bearer " + token}
-            client.Configuration.set_default(configuration)
+            self._get_iam_token()
 
         else:
             try:
@@ -77,10 +75,7 @@ class CodeEngineBackend:
                 self.code_engine_config['cluster'] = self.cluster
 
                 if self.iam_api_key:
-                    configuration = client.Configuration.get_default_copy()
-                    token = self._get_iam_token()
-                    configuration.api_key = {"authorization": "Bearer " + token}
-                    client.Configuration.set_default(configuration)
+                    self._get_iam_token()
 
             except Exception:
                 logger.debug('Loading incluster config')
@@ -105,17 +100,24 @@ class CodeEngineBackend:
         logger.info("{} - Region: {}".format(msg, self.region))
 
     def _get_iam_token(self):
-        """ Requests and IBM IAM token """
-        token = self.code_engine_config.get('token', None)
-        token_expiry_time = self.code_engine_config.get('token_expiry_time', None)
-        self.ibm_token_manager = IBMTokenManager(self.iam_api_key,
-                                                 'IAM', token,
-                                                 token_expiry_time)
+        """ Requests an IBM IAM token """
+        configuration = client.Configuration.get_default_copy()
+        if self.namespace and self.region:
+            configuration.host = self.cluster
+
+        if not self.ibm_token_manager:
+            token = self.code_engine_config.get('token', None)
+            token_expiry_time = self.code_engine_config.get('token_expiry_time', None)
+            self.ibm_token_manager = IBMTokenManager(self.iam_api_key,
+                                                     'IAM', token,
+                                                     token_expiry_time)
+
         token, token_expiry_time = self.ibm_token_manager.get_token()
         self.code_engine_config['token'] = token
         self.code_engine_config['token_expiry_time'] = token_expiry_time
 
-        return token
+        configuration.api_key = {"authorization": "Bearer " + token}
+        client.Configuration.set_default(configuration)
 
     def _format_jobdef_name(self, runtime_name, runtime_memory):
         runtime_name = runtime_name.replace('/', '--')
@@ -298,6 +300,12 @@ class CodeEngineBackend:
         """
         Clean all completed jobruns in the current executor
         """
+        if self.iam_api_key:
+            # try to refresh the token
+            self._get_iam_token()
+            self.custom_api = client.CustomObjectsApi()
+            self.core_api = client.CoreV1Api()
+
         if job_keys:
             for job_key in job_keys:
                 if job_key in self.jobs:
@@ -323,6 +331,12 @@ class CodeEngineBackend:
         Invoke -- return information about this invocation
         For array jobs only remote_invocator is allowed
         """
+        if self.iam_api_key:
+            # try to refresh the token
+            self._get_iam_token()
+            self.custom_api = client.CustomObjectsApi()
+            self.core_api = client.CoreV1Api()
+
         executor_id = job_payload['executor_id']
         job_id = job_payload['job_id']
 
@@ -364,7 +378,7 @@ class CodeEngineBackend:
                      .format(executor_id, job_id, total_calls, array_size))
 
         try:
-            res = self.custom_api.create_namespaced_custom_object(
+            self.custom_api.create_namespaced_custom_object(
                 group=ce_config.DEFAULT_GROUP,
                 version=ce_config.DEFAULT_VERSION,
                 namespace=self.namespace,
@@ -442,7 +456,7 @@ class CodeEngineBackend:
         container['resources']['requests']['cpu'] = str(self.code_engine_config['runtime_cpu'])
 
         try:
-            res = self.custom_api.delete_namespaced_custom_object(
+            self.custom_api.delete_namespaced_custom_object(
                 group=ce_config.DEFAULT_GROUP,
                 version=ce_config.DEFAULT_VERSION,
                 namespace=self.namespace,
@@ -453,7 +467,7 @@ class CodeEngineBackend:
             pass
 
         try:
-            res = self.custom_api.create_namespaced_custom_object(
+            self.custom_api.create_namespaced_custom_object(
                 group=ce_config.DEFAULT_GROUP,
                 version=ce_config.DEFAULT_VERSION,
                 namespace=self.namespace,

--- a/lithops/util/ibm_token_manager.py
+++ b/lithops/util/ibm_token_manager.py
@@ -37,7 +37,16 @@ class IBMTokenManager:
                          format(self._token_manager._expiry_time,
                                 self._get_token_minutes_diff()))
 
+    def _is_token_expired(self):
+        """
+        Checks if a token already expired
+        """
+        return self._get_token_minutes_diff() < 1
+
     def _get_token_minutes_diff(self):
+        """
+        Gets the remaining minutes in which the current token is valid
+        """
         expiry_time = self._token_manager._expiry_time
         return max(0, int((expiry_time - datetime.now(timezone.utc)).total_seconds() / 60.0))
 
@@ -48,12 +57,16 @@ class IBMTokenManager:
         token_data['token'] = self._token_manager._token
         token_data['token_expiry_time'] = self._token_manager._expiry_time.strftime('%Y-%m-%d %H:%M:%S.%f%z')
         dump_yaml_config(self._token_filename, token_data)
+        logger.debug("Token expiry time: {} - Minutes left: {}".
+                     format(self._token_manager._expiry_time,
+                            self._get_token_minutes_diff()))
 
     def get_token(self):
-        """ Gets a new token within a mutex block to prevent multiple threads
+        """
+        Gets a new token within a mutex block to prevent multiple threads
         requesting new tokens at the same time.
         """
-        if (self._token_manager._is_expired() or self._get_token_minutes_diff() < 1) \
+        if (self._token_manager._is_expired() or self._is_token_expired()) \
            and not is_lithops_worker():
             logger.debug("Token expired. Requesting new token".format(self.api_key_type))
             self._generate_new_token()

